### PR TITLE
Enhance MSAL Managed Identity to Bypass Cache When Claims Are Present

### LINKED ISSUE #4845: [Engineering task] Implement Claims API to Bypass Cache When Claims are Present in MSAL with Managed Identity
### Task type

Development

### Description

Currently, MSAL with Managed Identity does not expose any API claims API. With CAE (Continuous Access Evaluation) being enabled by default, we need to implement a mechanism to bypass the cache if claims are detected in the token request.

**Steps to Reproduce:**
Enable CAE by default in MSAL with Managed Identity.
Make a token request with claims present.

Observe that the cache is not bypassed, leading to potential stale token usage.

**Expected Behavior:**
When claims are present in the token request, the cache should be bypassed to ensure that the latest token is used, in line with CAE requirements.

### Solution

**Proposed Solution:**

1. Expose the claims API in MSAL for MI 
2. Expose Claims to MI Assertion Provider for FIC

### PR DESCRIPTION
**Pull request by gladjohn**
Fixes #4845

Currently, MSAL with Managed Identity does not expose any API for handling claims. With Continuous Access Evaluation (CAE) in MSI, it is necessary to implement a mechanism to bypass the cache if claims are detected in the token request.

**Steps to Reproduce:**

- Enable CAE by default in MSAL with Managed Identity.
- Make a token request with claims present.
- Observe that the cache is not bypassed, leading to potential stale token usage.

**Expected Behavior:**
When claims are present in the token request, the cache should be bypassed to ensure that the latest token is used, in line with CAE requirements.

**Proposed Solution:**

- Expose the claims API in MSAL for Managed Identity.
- Update the Managed Identity Assertion Provider to handle claims and ensure the cache is bypassed when claims are detected.

**Implementation Details:**

- Add a new API to MSAL for handling claims in Managed Identity.
- Modify the token acquisition logic to check for claims and bypass the cache if necessary.
- Ensure proper logging to indicate when the cache is bypassed due to claims being present.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Internal/Requests/ManagedIdentityAuthRequest.cs [hunk 1]
```diff
@@ -33,11 +33,11 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
...
             ILoggerAdapter logger = AuthenticationRequestParameters.RequestContext.Logger;
 
-            // Skip checking cache when force refresh is specified
-            if (_managedIdentityParameters.ForceRefresh)
+            // Skip checking cache when force refresh or claims is specified
+            if (_managedIdentityParameters.ForceRefresh || !string.IsNullOrEmpty(AuthenticationRequestParameters.Claims))
             {
                 AuthenticationRequestParameters.RequestContext.ApiEvent.CacheInfo = CacheRefreshReason.ForceRefreshOrClaims;
-                logger.Info("[ManagedIdentityRequest] Skipped looking for a cached access token because ForceRefresh was set.");
+                logger.Info("[ManagedIdentityRequest] Skipped looking for a cached access token because ForceRefresh or Claims were set.");
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/ManagedIdentityAuthRequest.cs:40]: Maybe it's worth knowing exactly why the cache was skipped?

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs [hunk 2]
```diff
@@ -62,6 +62,22 @@ public AcquireTokenForManagedIdentityParameterBuilder WithForceRefresh(bool forc
...
+        /// <summary>
+        /// Adds a claims challenge to the token request. The SDK will bypass the token cache when a claims challenge is specified.. Retry the 
+        /// token acquisition, and use this value in the <see cref="WithClaims(string)"/> method. A claims challenge typically arises when 
+        /// calling the protected downstream API, for example when the tenant administrator wants to revokes credentials. Apps are required 
+        /// to look for a 401 Unauthorized response from the protected api and to parse the WWW-Authenticate response header in order to 
+        /// extract the claims.See https://aka.ms/msal-net-claim-challenge for details.
+        /// </summary>
+        /// <param name="claims">A string with one or multiple claims.</param>
+        /// <returns>The builder to chain .With methods.</returns>
+        public AcquireTokenForManagedIdentityParameterBuilder WithClaims(string claims)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs:74]: What about client capabilities?

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs [hunk 3]
```diff
@@ -62,6 +62,22 @@ public AcquireTokenForManagedIdentityParameterBuilder WithForceRefresh(bool forc
             return this;
         }
 
+        /// <summary>
+        /// Adds a claims challenge to the token request. The SDK will bypass the token cache when a claims challenge is specified.. Retry the
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs:66]:
```suggestion
        /// Adds a claims challenge to the token request. The SDK will bypass the token cache when a claims challenge is specified. Retry the 
```

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs [hunk 4]
```diff
@@ -62,6 +62,22 @@ public AcquireTokenForManagedIdentityParameterBuilder WithForceRefresh(bool forc
             return this;
         }
 
+        /// <summary>
+        /// Adds a claims challenge to the token request. The SDK will bypass the token cache when a claims challenge is specified.. Retry the 
+        /// token acquisition, and use this value in the <see cref="WithClaims(string)"/> method. A claims challenge typically arises when 
+        /// calling the protected downstream API, for example when the tenant administrator wants to revokes credentials. Apps are required
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs:68]:
```suggestion
        /// calling the protected downstream API, for example when the tenant administrator revokes credentials. Apps are required 
```

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs [hunk 5]
```diff
@@ -62,6 +62,22 @@ public AcquireTokenForManagedIdentityParameterBuilder WithForceRefresh(bool forc
             return this;
         }
 
+        /// <summary>
+        /// Adds a claims challenge to the token request. The SDK will bypass the token cache when a claims challenge is specified.. Retry the 
+        /// token acquisition, and use this value in the <see cref="WithClaims(string)"/> method. A claims challenge typically arises when 
+        /// calling the protected downstream API, for example when the tenant administrator wants to revokes credentials. Apps are required 
+        /// to look for a 401 Unauthorized response from the protected api and to parse the WWW-Authenticate response header in order to 
+        /// extract the claims.See https://aka.ms/msal-net-claim-challenge for details.
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs:70]:
```suggestion
        /// extract the claims. See https://aka.ms/msal-net-claim-challenge for details.
```

File: tests/Microsoft.Identity.Test.Integration.netcore/HeadlessTests/ManagedIdentityTests.NetFwk.cs [hunk 6]
```diff
@@ -128,6 +128,76 @@ public async Task AcquireMSITokenAsync(MsiAzureResource azureResource, string us
             }
         }
 
+        [DataTestMethod]
+        [DataRow(MsiAzureResource.WebApp, "", DisplayName = "System_Identity_Web_App")]
+        [DataRow(MsiAzureResource.WebApp, UserAssignedClientID, UserAssignedIdentityId.ClientId, DisplayName = "ClientId_Web_App")]
+        [DataRow(MsiAzureResource.WebApp, UamiResourceId, UserAssignedIdentityId.ResourceId, DisplayName = "ResourceID_Web_App")]
+        [DataRow(MsiAzureResource.WebApp, UserAssignedObjectID, UserAssignedIdentityId.ObjectId, DisplayName = "ObjectID_Web_App")]
+        public async Task AcquireMSITokenWithClaimsAsync(MsiAzureResource azureResource, string userIdentity, UserAssignedIdentityId userAssignedIdentityId = UserAssignedIdentityId.None)
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netcore/HeadlessTests/ManagedIdentityTests.NetFwk.cs:136]: What is the value of this integration test? You are only testing MSAL internal behavior - ability to bypass the cache. I recommend you stick to unit testing, as integration tests are expensive.

File: tests/Microsoft.Identity.Test.Integration.netcore/HeadlessTests/ManagedIdentityTests.NetFwk.cs [hunk 7]
```diff
@@ -285,6 +355,109 @@ public async Task AcquireMsiToken_ExchangeForEstsToken_Successfully()
             }
         }
 
+        [TestMethod]
+        public async Task AcquireMsiToken_ExchangeForEstsTokenWithClaims_Successfully()
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netcore/HeadlessTests/ManagedIdentityTests.NetFwk.cs:359]: Same comment for this one - you are only bypassing the cache. Unit test is sufficient.

**Review by bgavrilMS — APPROVED** [review]

**Comment by bgavrilMS** [comment]
Why is it draft? Looks complete to me.

**Comment by gladjohn** [comment]
> Why is it draft? Looks complete to me.

marked it as ready for review and have added the new APIs as experimental @localden FYI

**Review by localden — APPROVED** [review]

**Pull request merged**